src/parallel/common/memory_management.cu

```diff
@@ -1,68 +1,68 @@
+    ERROR_CHECK(cudaMallocManaged((void **) &ca->Sy, sizeof(double) * ROWS * COLS));
+    ERROR_CHECK(cudaMallocManaged((void **) &ca->sources, sizeof(double) * ROWS * COLS));
+}
 
-	CUDASAFECALL(cudaMemcpy(d_write_head, h_ca.head, sizeof(double) * ROWS * COLS, cudaMemcpyHostToDevice));
+void allocateMemory(CA *&ca, double *&headsWrite)
+{
+    ERROR_CHECK(cudaMalloc((void **) &headsWrite, sizeof(double) * ROWS * COLS));
+    ERROR_CHECK(cudaMalloc((void **) &ca->heads, sizeof(double) * ROWS * COLS));
+    ERROR_CHECK(cudaMalloc((void **) &ca->K, sizeof(double) * ROWS * COLS));
+    ERROR_CHECK(cudaMalloc((void **) &ca->Sy, sizeof(double) * ROWS * COLS));
+    ERROR_CHECK(cudaMalloc((void **) &ca->sources, sizeof(double) * ROWS * COLS));
+}
 
+void copyDataFromCpuToGpu(CA *&h_ca, CA *&d_ca)
+{
+    ERROR_CHECK(cudaMemcpy(d_ca->heads, h_ca->heads, sizeof(double) * ROWS * COLS, cudaMemcpyHostToDevice));
+    ERROR_CHECK(cudaMemcpy(d_ca->K, h_ca->K, sizeof(double) * ROWS * COLS, cudaMemcpyHostToDevice));
+    ERROR_CHECK(cudaMemcpy(d_ca->Sy, h_ca->Sy, sizeof(double) * ROWS * COLS, cudaMemcpyHostToDevice));
+    ERROR_CHECK(cudaMemcpy(d_ca->sources, h_ca->sources, sizeof(double) * ROWS * COLS, cudaMemcpyHostToDevice));
 }
 
-void init_host_ca() {
-	h_ca.head = new double[ROWS * COLS]();
-	h_ca.Sy = new double[ROWS * COLS]();
-	h_ca.K = new double[ROWS * COLS]();
-	h_ca.Source = new double[ROWS * COLS]();
+void initializeCA(CA *&ca)
+{
+    int wellsRows[] = {WELLS_Y};
+    int wellsCols[] = {WELLS_X};
+    double wellsQW[] = {WELLS_QW};
 
-	for (int i = 0; i < ROWS; i++)
-		for (int j = 0; j < COLS; j++) {
-			h_ca.head[i * ROWS + j] = headFixed;
-			if (j == COLS - 1) {
-				h_ca.head[i * ROWS + j] = headCalculated;
-			}
-			h_ca.Sy[i * ROWS + j] = Syinitial;
-			h_ca.K[i * ROWS + j] = Kinitial;
-			h_ca.Source[i * ROWS + j] = 0;
-		}
+    for (int i{}; i < ROWS; ++i)
+        for (int j{}; j < COLS; ++j)
+        {
+            ca->heads[i * ROWS + j] = headFixed;
+            if (j == COLS - 1)
+            {
+                ca->heads[i * ROWS + j] = headCalculated;
+            }
+            ca->Sy[i * ROWS + j] = Syinitial;
+            ca->K[i * ROWS + j] = Kinitial;
+            ca->sources[i * ROWS + j] = 0;
+        }
 
-	int x,y;
-	double source;
-	for(int i = 0; i < NUMBER_OF_WELLS; i++){
-		x = wellsRows[i];
-		y = wellsCols[i];
-		source = wellsQW[i];
-		h_ca.Source[y *ROWS + x] = source;
-	}
+    for (int i{}; i < NUMBER_OF_WELLS; ++i)
+    {
+        int x = wellsRows[i];
+        int y = wellsCols[i];
+        ca->sources[y * ROWS + x] = wellsQW[i];
+    }
 }
 
-void copy_data_from_GPU_to_CPU() {
-	CUDASAFECALL(cudaMemcpy(h_ca.head, d_write_head, sizeof(double) * ROWS * COLS, cudaMemcpyDeviceToHost));
+void copyDataFromGpuToCpu(CA *&h_ca, CA *&d_ca)
+{
+    ERROR_CHECK(cudaMemcpy(h_ca->heads, d_ca->heads, sizeof(double) * ROWS * COLS, cudaMemcpyDeviceToHost));
 }
 
+void freeAllocatedMemory(CA *&d_ca, double *&headsWrite)
```

Comment:
Brakuje zwolnienia zaalokowaniej pamięci dla CA na hoście

Reply:
wywale to z unified, nie chce mi się dodawać jak to w ogole nie jest potrzebne

Reply:
albo nie zostawie, bo jest w obu, ale dodawć mi się nie chce zwalniania wszystkiego